doc/design/state-machine.md

```diff
@@ -0,0 +1,185 @@
+
+Common cell life cycles:
+
+Free -> Used -> Free (no preemption involved)
+
+Free -> Used -> Reserving -> Reserved -> Used -> Free (preemption involved)
+
+### Events
+
+Note: “cell view” in the below operations represents the in-memory data structures for scheduling, e.g., free cell list, cell bindings, cell priorities. Allocating / releasing a cell in the cell view will modify the free cell list (and split / merge the cell), create / destroy cell bindings, and set / reset the cell priority. These changes are immediately visible to the subsequent scheduling cycles.
+
+__e<sub>c</sub>__:
+
+Condition: scheduler crashes and restarts.
+
+Operation: none.
+
+Note that all states are volatile; the Free and Used states are derived from the AG state machine.
+
+__e<sub>0</sub>__:
+
+Condition: triggered by AG from Pending to Allocated (e<sub>0</sub> in AG state machine).
+
+Operation: allocate the cell to the AG in the cell view.
+
+__e<sub>1</sub>__:
+
+Condition: the Allocated AG is deleted (e<sub>3</sub> in AG state machine).
+
+Operation: release the cell in the cell view.
+
+__e<sub>2</sub>__:
+
+Condition: triggered by AG from Pending to Preempting (e<sub>1</sub> in AG state machine).
+
+Operation: set the Allocated AG on this cell to Being preempted state (e<sub>6</sub> in AG state machine); release the cell from the cell view, and then allocate it to the preemptor AG in the cell view.
+
+__e<sub>3</sub>__:
+
+Condition: triggered by AG from Pending to Preempting (e<sub>1</sub> in AG state machine).
+
+Operation: set the original preemptor AG to Pending state (e<sub>5</sub> in AG state machine); release the cell from the cell view, and then allocate it to the new preemptor.
+
+__e<sub>4</sub>__:
+
+Condition: triggered by AG from Preempting to Pending (e<sub>5</sub> in AG state machine) or from Preempting to Deleted (e7 in AG state machine).
+
+Operation: release the cell from the cell view, and then allocate it to the running AG (i.e., the preemption victim).
+
+__e<sub>5</sub>__:
+
+Condition: triggered by AG from Pending to Preempting (e<sub>1</sub> in AG state machine).
+
+Operation: allocate the cell to the preemptor in the cell view.
+
+__e<sub>6</sub>__:
+
+Condition: triggered by AG from Pending to Preempting (e<sub>1</sub> in AG state machine).
+
+Operation: release the cell from the cell view, and then allocate it to the new preemptor, and set the original preemptor AG from Preempting to Pending (e<sub>5</sub> in AG state machine).
+
+__e<sub>7</sub>__:
+
+Condition: triggered by AG from Preempting to Pending (e<sub>5</sub> in AG state machine) or from Preempting to Completed (e<sub>7</sub> in AG state machine).
+
+Operation: release the cell from the cell view.
+
+__e<sub>8</sub>__:
+
+Condition: triggered by AG from Preempting to Allocated (e<sub>2</sub> in AG state machine).
+
+Operation: none.
+
+__e<sub>9</sub>__:
+
+Condition: the running pod is preempted.
```

Comment:
We are doing gang preemption, so this will change victim AG states?

Reply:
This will not change victim AG state. Victim AG transitions to Deleted only when all pods are deleted (see AG e4)